homeassistant/components/combined_energy/sensor.py

```diff
@@ -0,0 +1,270 @@
+        if device_readings := self.device_readings:
+            return getattr(device_readings, self.entity_description.key)
+        return None
+
+    @property
+    def available(self) -> bool:
+        """Indicate if the entity is available."""
+        return self._raw_value is not None
+
+    @abstractmethod
+    def _to_native_value(self, raw_value):
+        """Convert non-none raw value into usable sensor value."""
+
+    @property
+    def native_value(self) -> int | float | None:
+        """Return the state of the sensor."""
+        if (raw_value := self._raw_value) is not None:
+            return self._to_native_value(raw_value)
+        return None
+
+
+class GenericSensor(CombinedEnergyReadingsSensor):
+    """Sensor that returns the last value of a sequence of readings."""
+
+    def _to_native_value(self, raw_value):
+        """Convert non-none raw value into usable sensor value."""
+        if isinstance(raw_value, Sequence):
+            raw_value = raw_value[-1]
+        return round(raw_value, self.native_value_rounding)
+
+
+class EnergySensor(CombinedEnergyReadingsSensor):
+    """Sensor for energy readings."""
+
+    @property
+    def last_reset(self) -> datetime | None:
+        """Last time the data was reset."""
+        if device_readings := self.device_readings:
+            return device_readings.range_start
+        return None
+
+    def _to_native_value(self, raw_value):
+        """Convert non-none raw value into usable sensor value."""
+        value = sum(raw_value)
+        return round(value, self.native_value_rounding)
+
+
+class PowerSensor(CombinedEnergyReadingsSensor):
+    """Sensor for power readings."""
+
+    def _to_native_value(self, raw_value):
+        """Convert non-none raw value into usable sensor value."""
+        return round(raw_value, self.native_value_rounding)
+
+
+class PowerFactorSensor(CombinedEnergyReadingsSensor):
+    """Sensor for power factor readings."""
+
+    native_value_rounding = 1
+
+    def _to_native_value(self, raw_value):
+        """Convert non-none raw value into usable sensor value."""
+        # The API expresses the power factor as a fraction convert to %
+        return round(raw_value[-1] * 100, self.native_value_rounding)
+
+
+class WaterVolumeSensor(CombinedEnergyReadingsSensor):
+    """Sensor for water volume readings."""
+
+    def _to_native_value(self, raw_value):
+        """Convert non-none raw value into usable sensor value."""
+        return int(round(raw_value[-1], 0))
+
+
+# Map of common device classes to specific sensor types
+SENSOR_TYPE_MAP: dict[
+    SensorDeviceClass | str | None, type[CombinedEnergyReadingsSensor]
+] = {
+    SensorDeviceClass.ENERGY: EnergySensor,
+    SensorDeviceClass.POWER: PowerSensor,
+    SensorDeviceClass.WATER: WaterVolumeSensor,
+    SensorDeviceClass.POWER_FACTOR: PowerFactorSensor,
+    None: GenericSensor,
+}
+
+
+class CombinedEnergyReadingsSensorFactory:
```

Comment:
I've never seen someone build a factory for this task and I'm not sure how the maintainers will feel about it.

Reply:
A factory makes sense when devices need to be enumerated from the combined energy API. They are not known before this process is completed.